### DISCUSSION TITLE
Powerful New Page Method `insert_htmlbox`

### DISCUSSION THREAD

**Discussion by JorjMcKie** [pymupdf/PyMuPDF#2914] [Announcements]
Many have waited for this functionality:
Method [Page.insert_htmlbox](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_htmlbox) writes HTML/CSS-enriched text into a rectangle of a PDF page.
It has many similarities with methods [Page.insert_textbox](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_textbox) and [TextWriter.fill_textbox](https://pymupdf.readthedocs.io/en/latest/textwriter.html#TextWriter.fill_textbox).

More important are the **_significant_** differences:
* Support of **_HTML tags_** and **_CSS styling_** instructions being part of the text: all syntax elements supported by [Story](https://pymupdf.readthedocs.io/en/latest/story-class.html#Story) are possible. As a consequence, all of the following is supported:
    - multiple fonts
    - user-supplied fonts, auto-fallback fonts
    - multiple font sizes and text colors
    - embedded hyperlinks
    - images
    - tables
    - multiple languages with automatic font selection
    - automatic **_"text shaping"_** (using [HarfBuzz](https://github.com/harfbuzz/harfbuzz)), finally yielding correct output of **_Hindi, Bengali, Tamil, Sanskrit_** etc. - text using complex scripting systems like Devanagari, right-to-left languages and many more.
* No more worries whether the boundary box is large enough: The content will automatically be **_scaled down to fit_**. The extent of acceptable down scaling can be specified.
    - As before: if no downscaling was necessary, a float is returned indicating the unused part of the rectangle.
* Content can be rotated by multiples of 90° as with the other methods.
* Text alignment and other effects, including font specification can be (and must be) achieved via HTML or CSS syntax - there are no method parameters for this.
* As an alternative to a text string, a Story object can be given. This allows for all the preparative adjustments achievable by manipulating the [Xml](https://pymupdf.readthedocs.io/en/latest/xml-class.html#Xml) object [Story.body](https://pymupdf.readthedocs.io/en/latest/story-class.html#Story.body) prior to executing `page.insert_htmlbox(bbox, story)`.

Please refer to a collection of simple examples in the [HOW TO GUIDE](https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-fill-a-box-with-html-text).

**Comment by ericpias** [comment]
I am very interested in using insert_htmlbox.  I upgraded to 1.23.8 and fitz.__version__ now reports 1.23.8.  The docs say that page.insert_htmlbox is new for 1.23.8.  However, the insert_htmlbox is not available on page in 1.23.8.  Any ideas?

**Reply by JorjMcKie** [reply]
This method is available for the new, "rebased" version of PyMuPDF.
For a short period of time, you must use the import statement `import fitz_new as fitz`. During the course of this month, we will enter phase 2 and switch versions, which will make rebased the default fitz and keep the classic version available as `fitz_old` as a fallback.

**Reply by ericpias** [reply]
Thanks.

**Comment by lexxwork** [comment]
How to remove/clear/erase all content made by `insert_htmlbox` leaving background image/color int the defined rect?
Maybe i should ask this question on mupdf page?

**Reply by JorjMcKie** [reply]
You can as well ask here 😎.
Images, links and text as inserted by this method can be removed via redaction annotations as usual.
Vector graphics however (like background colors, table grid lines) cannot be removed yet.
Making vector graphics removal also subject to redaction annotations, is work underway in MuPDF. Once supported there, it will subsequently be made available in PyMuPDF.

**Reply by lexxwork** [reply]
great! looks like it works, at least for text.
here is next issue related to the complete erasing bangla text
here is the result of my test
![image](https://github.com/pymupdf/PyMuPDF/assets/7888697/4e6d504d-e32c-44ee-b1e9-c916a51a4bfb)
you ma see that erased area has artifacts
for removing text i used two methods:
1. regular 
```python
page.add_redact_annot(block_clip1)
page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_NONE)
```
2. my custom function _**erase_text_in_rect**_ that works well with bangla text after fitz.TextWriter
you may find it in 
[hindi-test.zip](https://github.com/pymupdf/PyMuPDF/files/14215010/hindi-test.zip)

can you suggest another approach to completely remove text?
* updated zip file

**Reply by JorjMcKie** [reply]
The zip file is empty!
But this seems to be a bug in MuPDF. I will open one in its issue tracker.

**Reply by JorjMcKie** [reply]
[redact.zip](https://github.com/pymupdf/PyMuPDF/files/14215062/redact.zip)

The ZIP contains `input.pdf` with text containing Devanagari glyphs among others and is wrapped by a redaction annotation.
When applying redactions, the text will not get completely removed: some text artefacts are remaining.
A reproducer C-source and the output PDF are also in the ZIP.

MuPDF issue [here](https://bugs.ghostscript.com/show_bug.cgi?id=707560).

**Reply by lexxwork** [reply]
Thank you for your so fast reaction!
I have updated my archive, please look at `erase_text_in_rect` function. It must work but not for this case.

**Reply by lexxwork** [reply]
seems they have fixed it
![image](https://github.com/pymupdf/PyMuPDF/assets/7888697/f6a5328a-abc0-4a48-9afa-d86321be3627)
https://bugs.ghostscript.com/show_bug.cgi?id=707560#c2

**Reply by JorjMcKie** [reply]
Yes, I know they have 😎.

**Reply by lexxwork** [reply]
@JorjMcKie, i did update to latest version PyMuPDF-1.23.26, the issue is still there, so it means you did not update mupdf?

**Reply by JorjMcKie** [reply]
It is in the next version ...

**Reply by JB2001216** [reply]
Hi @JorjMcKie , could you please let me know when the next version will be published with this issue resolved?
